pkg/queue/handlers/http_request.go

```diff
@@ -0,0 +1,258 @@
+	span.SetTag("task.queue", task.Queue)
+	span.SetTag("task.type", task.Type)
+	span.SetTag("task.spec", string(task.Spec))
+
+	logrus := logrus.WithField("type", task.Type).WithField("id", task.ID)
+
+	logrus.Debug("starting the HTTP request task...")
+
+	var progress HTTPRequestProgress
+	defer func() {
+		// we check for errSerializingHearbeat so we don't cause
+		// a recursion call
+		if err == nil || err == ErrSerializingHearbeat {
+			return
+		}
+		message := err.Error()
+		progress.ErrorMessage = &message
+		_ = sendHTTPRequestProgress(progress, heartbeats)
+	}()
+
+	var spec HTTPRequestTaskSpec
+	err = json.Unmarshal(task.Spec, &spec)
+	if err != nil {
+		return err
+	}
+
+	progress.Stage = RequestPreparing
+	err = sendHTTPRequestProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	var payload io.Reader
+	if spec.RequestBody != "" {
+		payload = strings.NewReader(spec.RequestBody)
+	}
+
+	req, err := http.NewRequest(spec.Method, spec.URL, payload)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("User-Agent", "Contiamo Hub HTTP Request Task")
+
+	for name, value := range spec.RequestHeaders {
+		req.Header.Add(name, value)
+	}
+
+	if spec.Authorized {
+		token, err := h.tokenCreator.Create("httpRequestTask")
+		if err != nil {
+			return err
+		}
+
+		req.Header.Add(h.tokenHeaderName, token)
+	}
+
+	err = opentracing.GlobalTracer().Inject(
+		span.Context(),
+		opentracing.HTTPHeaders,
+		opentracing.HTTPHeadersCarrier(req.Header),
+	)
+	if err != nil {
+		otext.Error.Set(span, true)
+		span.SetTag("tracing.inject.err", err.Error())
+		err = nil
+	}
+
+	progress.Stage = RequestPending
+	err = sendHTTPRequestProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	defer func() {
+		duration := time.Since(now).Milliseconds()
+		progress.Duration = &duration
+		err := sendHTTPRequestProgress(progress, heartbeats)
+		if err != nil {
+			logrus.Error(err)
+		}
+	}()
+
+	resp, err := h.client.Do(req.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	contentType := resp.Header.Get("content-type")
+	if !strings.Contains(contentType, "application/json") {
+		return fmt.Errorf(
+			"unexpected response content type, expected `application/json`, got `%s`",
+			contentType,
+		)
+	}
+
+	progress.Stage = RequestResponse
+	progress.ReturnedStatus = &resp.StatusCode
+	err = sendHTTPRequestProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	// the task would time out if the heartbeat was not sent for 30 seconds
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				err := sendHTTPRequestProgress(progress, heartbeats)
+				if err != nil {
+					logrus.Error(err)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(resp.Body)
+	for decoder.More() {
+		err = ctx.Err()
+		if err != nil {
+			return err
+		}
+		var m json.RawMessage
```

Comment:
that makes no sense to me. Decoding into a RawMessage means copying the data and nothing more. Perhaps it checks if its valid json, but nothing more. Why not just ignore the content type and pass through the body untouched?

Reply:
got it, only json is supported.